### PR TITLE
Add evaluate_guards option to DynamicShapesConfig

### PR DESCRIPTION
**Pull request by laithsakka**
evaluate_guards is A debug mode to detect and fail if Dynamo ever specializes a dynamic shape by
guarding on it. When True, dynamic shape guards are not dropped from Dynamo.
And a failure will be triggered if recompilation ever happens due to that.

Enabling this allow observing the dynamic shapes guards in the tl-parse artifact.

One exception that evaluate_guards wont verify is counter 0/1 specializations. 
Backed dynamic shapes by default will assume sizes >=2. we will not check nor fail 
if that guard fails for 0/1 inputs sizes.

### CODE REVIEW

**Comment by hmellor** [comment]
> Note : Only last(top) commit belongs to this PR

Which PR is this built on top of?

**Comment by laithsakka** [comment]
> > Note : Only last(top) commit belongs to this PR
> 
> Which PR is this built on top of?

https://github.com/vllm-project/vllm/pull/26199
and

https://github.com/vllm-project/vllm/pull/25110

**Comment by hmellor** [comment]
Ok, should we wait for those to merge before reviewing this one?

File: vllm/config/compilation.py [hunk 1]
```diff
@@ -208,6 +208,14 @@ class DynamicShapesConfig:
       backed/unbacked.
     """
 
+    evaluate_guards: bool = False
```
Comment (zou3519) [inline vllm/config/compilation.py:347]: I think I want to get into a state where this is always on during vLLM warmup, and then optionally during runtime. This involves making sure the guards are actually correct (maybe this requires unbacked). Not for this PR, but we should do this in a follow-up.

File: vllm/compilation/wrapper.py [hunk 2]
```diff
@@ -41,10 +43,15 @@ def __init__(self):
...
         if isinstance(backend, str) and backend == "inductor":
             options = vllm_config.compilation_config.inductor_compile_config
-
+        self.first_compile = True
         if mode != CompilationMode.STOCK_TORCH_COMPILE:
             # Drop all the guards.
-            options["guard_filter_fn"] = lambda x: [False for _ in x]
+            if vllm_config.compilation_config.dynamic_shapes_config.evaluate_guards:
+                options["guard_filter_fn"] = lambda x: [
+                    entry.guard_type == "SHAPE_ENV" for entry in x
```
Comment (zou3519) [inline vllm/compilation/wrapper.py:153]: this seems brittle..
can we add a test in pytorch for this or something? "if you change how the dynamic shape guards look then you need to change vLLM"

File: tests/compile/test_dynamic_shapes_compilation.py [hunk 3]
```diff
@@ -110,36 +115,136 @@ def test_dynamic_shapes_compilation(monkeypatch, model_name):
...
+                model(input_5)
+                # If we get here, no guard violation occurred
+                # This is a TEST FAILURE - evaluate_guards should have caused a failure
+                pytest.fail(
+                    f"{config_desc}: Expected guard violation did "
+                    f"not occur! evaluate_guards=True should fail "
+                    f"when shape changes from 10 to 5, but the "
+                    f"model ran successfully without error."
+                )
+            except Exception as e:
```
Comment (zou3519) [inline tests/compile/test_dynamic_shapes_compilation.py]: is there a more specific exception we can add?

File: tests/compile/test_dynamic_shapes_compilation.py [hunk 4]
```diff
@@ -110,36 +115,136 @@ def test_dynamic_shapes_compilation(monkeypatch, model_name):
...
+                if (
+                    "recompile" in str(e).lower()
+                    and "fail_on_recompile" in str(e).lower()
+                ):
+                    print(f"✅ {config_desc}: Recompile occurred (expected behavior)")
+                    print("   Recompiles are allowed when evaluate_guards=False")
+                else:
+                    print(f"❌ {config_desc}: Unexpected failure")
+                    print(f"   Error: {e}")
+                    raise
```
Comment (zou3519) [inline tests/compile/test_dynamic_shapes_compilation.py]: can you add a test that we dropped 0/1 specialization in backed mode?

File: tests/compile/test_dynamic_shapes_compilation.py [hunk 5]
```diff
@@ -110,36 +115,136 @@ def test_dynamic_shapes_compilation(monkeypatch, model_name):
...
+                # Expected failure - guard was violated
+                error_msg = str(e)
+                if "guard" in error_msg.lower() or "recompile" in error_msg.lower():
+                    print(f"✅ {config_desc}: Expected failure due to guard violation")
+                    print(f"   Error (truncated): {error_msg[:150]}")
+                else:
+                    # Unexpected error type
+                    print(f"❌ {config_desc}: Unexpected error type")
+                    print(f"   Error: {e}")
+                    raise
```
Comment (zou3519) [inline tests/compile/test_dynamic_shapes_compilation.py]: tests aren't really supposed to print. why not pytest.fail for the failure case?

**Review by zou3519 — COMMENTED** [review]
seems reasonable, just had testing qs

File: vllm/compilation/wrapper.py (hunk 2 above)
Comment (laithsakka) [reply vllm/compilation/wrapper.py:153]: yeh we shall if we dont have any. I will file issue and assign to me.

File: tests/compile/test_dynamic_shapes_compilation.py [hunk 6]
```diff
@@ -86,3 +96,100 @@ def test_dynamic_shapes_compilation(
...
+        )
+    )
+
+    def test(model_class, input1, input2, is_01_specialization=False):
+        with torch.no_grad(), use_vllm_config(vllm_config):
+            model = model_class(vllm_config=vllm_config).cuda()
+
+            model(input1)
+
+            if evaluate_guards and not is_01_specialization:
```
Comment (ProExpertProg) [inline tests/compile/test_dynamic_shapes_compilation.py]: Can we use `with pytest.raises()`?

File: vllm/compilation/backends.py [hunk 7]
```diff
@@ -752,6 +754,27 @@ def __call__(
...
+            # torch.compile will specialize for 0/1 inputs or otherwise guards that
+            # shape is >= 2. This is because it's really hard not to hit a check
+            # against 0/1. When we evaluate shape guards, we exclude checking those
+            # guards (We would fail always otherwise).
+
+            # We avoid that by updating the ranges of backed sizes when the min is
+            # 2 for any, we assume it's 0.
+            for s, r in fake_mode.shape_env.var_to_range.items():
+                if r.lower == 2:
+                    fake_mode.shape_env.var_to_range[s] = ValueRanges(0, r.upper)
```
Comment (ProExpertProg) [inline vllm/compilation/backends.py:747]: Can you import `ValueRanges` locally as it's an internal?

File: vllm/compilation/backends.py [hunk 8]
```diff
@@ -752,6 +754,27 @@ def __call__(
             self.split_gm, submod_names_to_compile, self.vllm_config, self
         ).run(*example_inputs)
 
+        from torch._guards import detect_fake_mode
+
+        fake_mode = detect_fake_mode()
+
+        if (
```
Comment (ProExpertProg) [inline vllm/compilation/backends.py:730]: Why do we only do this with `evaluate_guards` on?

Comment (ProExpertProg) [reply vllm/compilation/backends.py:730]: Oh is it because otherwise we drop all guards anyway?

File: vllm/config/compilation.py (hunk 1 above)
Comment (laithsakka) [reply vllm/config/compilation.py:347]: using unbacked basically just avoids the need of this, once we address the perf gp

File: vllm/compilation/wrapper.py (hunk 2 above)
Comment (laithsakka) [reply vllm/compilation/wrapper.py:153]: actually vllm testd would fail I also found a an internal test that checks it         def test_symbool_guards(

File: vllm/compilation/backends.py (hunk 8 above)
Comment (laithsakka) [reply vllm/compilation/backends.py:730]: yes

Comment (laithsakka) [reply vllm/compilation/backends.py:730]: also for backed size oblivious we do no t want to drop 0/1 guards

**Review by laithsakka — COMMENTED** [review]
rebase and address all comments

File: tests/compile/test_dynamic_shapes_compilation.py [hunk 9]
```diff
@@ -86,3 +97,118 @@ def test_dynamic_shapes_compilation(
...
+                    assert (
+                        "GuardManager check failed" in error_msg
+                        or "Detected recompile when torch.compile stance" in error_msg
+                    ), error_msg
+
+            else:
+                model(input2)
+
+    # test(ModelWithSizeCheck, torch.randn(20, 10).cuda(), torch.randn(5, 10).cuda())
+    # test(ModelWithSizeCheck, torch.randn(5, 10).cuda(), torch.randn(20, 10).cuda())
```
Comment (zou3519) [inline tests/compile/test_dynamic_shapes_compilation.py]: why the comments? did you mean to enable this?

File: tests/compile/test_dynamic_shapes_compilation.py [hunk 10]
```diff
@@ -86,3 +97,118 @@ def test_dynamic_shapes_compilation(
...
+                try:
+                    model(input2)
+                    raise RuntimeError("expected guard violation to occur")
+                except RuntimeError as e:
+                    # Expected failure - guard was violated
+                    error_msg = str(e)
+                    assert (
+                        "GuardManager check failed" in error_msg
+                        or "Detected recompile when torch.compile stance" in error_msg
+                    ), error_msg
```
Comment (zou3519) [inline tests/compile/test_dynamic_shapes_compilation.py]: nit: use pytest.raises? https://docs.pytest.org/en/7.1.x/how-to/assert.html#assertions-about-expected-exceptions

File: vllm/compilation/wrapper.py (hunk 2 above)
Comment (zou3519) [reply vllm/compilation/wrapper.py:153]: sgtm

File: vllm/config/compilation.py [hunk 11]
```diff
@@ -227,7 +227,18 @@ class DynamicShapesConfig:
...
     """
 
-    # TODO add a debug mode to fail
+    evaluate_guards: bool = False
+    """
+    A debug mode to detect and fail if Dynamo ever specializes a dynamic shape by
+    guarding on it. When True, dynamic shape guards are not dropped from dynamo.
+    And a failure will be triggered if a recompilation ever happens due to that.
+    This mode requires VLLM_USE_BYTECODE_HOOK to be 0.
+    Enabling this allow observing the dynamic shapes guards in the tl-parse
```
Comment (zou3519) [inline vllm/config/compilation.py]: nit: tlparse is one word?

**Review by zou3519 — APPROVED** [review]
LGTM, minor nits, might need a rebase too

File: tests/compile/test_dynamic_shapes_compilation.py (hunk 9 above)
Comment (laithsakka) [reply tests/compile/test_dynamic_shapes_compilation.py]: oh yes lol

File: vllm/compilation/wrapper.py [hunk 12]
```diff
@@ -107,23 +109,53 @@ def __init__(self):
...
+                    # is picked up.
+                    assert ds_type != DynamicShapesType.BACKED, (
+                        "evaluate_guards for backed shapes requires "
+                        "VLLM_USE_AOT_COMPILE=False. "
+                    )
+
+                assert not envs.VLLM_USE_BYTECODE_HOOK, (
+                    "compilation_config.dynamic_shapes_config.evaluate_guards "
+                    "requires VLLM_USE_BYTECODE_HOOK=0. "
+                )
```
Comment (ProExpertProg) [inline vllm/compilation/wrapper.py]: Duplicate?

File: vllm/compilation/wrapper.py [hunk 13]
```diff
@@ -107,23 +109,53 @@ def __init__(self):
...
-                    "UNBACKED dynamic shapes require VLLM_USE_BYTECODE_HOOK=0. "
-                )
+            # reason is that bytecode does torch._dynamo.eval_frame.
+            # remove_from_cache(self.original_code_object()) to force a new
+            # re-compilation. And if we use
+            # compiled_ptr = self.check_invariants_and_forward
+            # it will reset all entries.
+            assert envs.VLLM_USE_BYTECODE_HOOK, (
+                "UNBACKED dynamic shapes requires VLLM_USE_BYTECODE_HOOK=0. "
+            )
```
Comment (ProExpertProg) [inline vllm/compilation/wrapper.py]: assert is the inverse of the message

**Review by ProExpertProg — COMMENTED** [review]
Two minor comments:

File: docker/Dockerfile [hunk 14]
```diff
@@ -188,15 +188,14 @@ ARG nvcc_threads=8
 ENV NVCC_THREADS=$nvcc_threads
 
 ARG USE_SCCACHE
-ARG SCCACHE_DOWNLOAD_URL=https://github.com/mozilla/sccache/releases/download/v0.8.1/sccache-v0.8.1-x86_64-unknown-linux-musl.tar.gz
```
Comment (laithsakka) [inline docker/Dockerfile:191]: what is this lol
